### DISCUSSION TITLE
Necessity of refactor of UnitCell class

### DISCUSSION THREAD

**Discussion by kirk0830** [deepmodeling/abacus-develop#3024] [Ideas]
Presently UnitCell brings TOO MANY troubles when coding unit test, there are always tons of unpleasant `undefined reference to xxx` error messages pop on screen when compile. UnitCell is needed to be refactored mainly for development convienence.
Is it nessary and will be marked as high priority in the near future?

**Comment by maki49** [comment]
I also want some refactor. It is because I want to move Symmetry into UnitCell to be used in module Relax for #2901, but now Symmetry itself depends on UnitCell...

My idea is to divide UnitCell into several sub-modules:
 - Atoms
 - Lattice
 - Magnetism
 - Symmetry (depending on Atoms and Lattice rather than the entire UnitCell)

**Reply by kirk0830** [reply]
UnitCell is quite a heavy class but, have to say it is because structural info. itself has plenty of aspects. Refactoring UnitCell would be a topic neeeded to be considered cautiously... An initial outcome will be a careful design. We need more developers participate in this discussion...
For my opionion UnitCell is an abstract concept, 'cell' would be a carrier storing Bravis lattice information, atoms as contents of cell but... one should consider carefully about where to store atom-wise info. Present implementation collects almost everything but too centralized to make waste memory and make code complicated too much.
I mean there should be an answer to question about relationship between classes, parallel, or include or being included, case-specific or universal, ...
Another topic would be mpi distribution, but before writing new codes, it is fairly needed to tell all developers "where to find something", "what is needed to distrbute" and vice versa.
